### PR TITLE
Implement non-mod.rs mod statements

### PR DESCRIPTION
**Pull request by cramertj**
Fixes https://github.com/rust-lang/rust/issues/45385, cc https://github.com/rust-lang/rust/issues/44660

This will fail tidy right now because it doesn't recognize my UI tests as feature-gate tests. However, I'm not sure if compile-fail will work out either because compile-fail usually requires there to be error patterns in the top-level file, which isn't possible with this feature. What's the recommended way to handle this?

### CODE REVIEW

**Comment by zackmdavis** [comment]
> because compile-fail usually requires there to be error patterns in the top-level file, which isn't possible with this feature. What's the recommended way to handle this?

There's a special `rustc_error` attribute (gated by `rustc_attrs`) that you can [use to make the test artificially fail](https://github.com/rust-lang/rust/blob/master/src/test/compile-fail/feature-gate-fn_must_use.rs#L24-L31).

**Comment by est31** [comment]
> This will fail tidy right now because it doesn't recognize my UI tests as feature-gate tests.

You can also add the ui directory to the places where tidy searches for feature gate tests. Look into `src/tools/tidy/src/features.rs`. Then you need to name the files after the format that tidy recognizes.

**Comment by nikomatsakis** [comment]
> There's a special rustc_error attribute (gated by rustc_attrs) that you can use to make the test artificially fail.

You can use `// must-compile-successfully`, which I believe is now the preferred approach.

**Comment by nikomatsakis** [comment]
That said, what @est31 said sounds better. =)

File: src/libsyntax/parse/mod.rs [hunk 1]
```diff
@@ -86,9 +90,13 @@ pub struct Directory {
 
 #[derive(Copy, Clone)]
 pub enum DirectoryOwnership {
-    Owned,
+    Owned {
+        // None if `mod.rs`, `Some(path)` if we're in
+        // `foo.rs` and there's a corresponding `foo/...`
```
Comment (nikomatsakis) [inline src/libsyntax/parse/mod.rs]: This comment makes it sound like we check for the existence of such a directory, but we don't right? Maybe "`Some("foo")` if we're in `foo.rs`?

**Comment by est31** [comment]
btw this PR motivated me to move over some of the currently existing feature gate tests: #46532

File: src/libsyntax/feature_gate.rs [hunk 2]
```diff
@@ -1295,6 +1298,16 @@ fn contains_novel_literal(item: &ast::MetaItem) -> bool {
     }
 }
 
+impl<'a> PostExpansionVisitor<'a> {
+    fn whole_crate_feature_gates(&mut self) {
+        for &span in &*self.context.parse_sess.non_modrs_mods.borrow() {
+            gate_feature_post!(
+                &self, non_modrs_mods, span,
+                "mod declarations in non-mod.rs files are experimental");
```
Comment (nikomatsakis) [inline src/libsyntax/feature_gate.rs]: maybe we want to add some extra notes here, given that this is a common source of confusion? Something like:

> on stable builds, rename `foo.rs` to `foo/mod.rs` to enable submodules

not sure how hard that would be to do

**Review by nikomatsakis — COMMENTED** [review]
looks pretty good to me -- I'd like to see a `ui` test for what happens when there is a `mod foo` with no matching file (what kind of error do we give)

**Comment by cramertj** [comment]
@nikomatsakis 
> I'd like to see a ui test 

I've added one, but currently it just suggests to try creating the file, and once the file is created it will give you the stability error.

File: src/test/ui/invalid-module-declaration/invalid-module-declaration.stderr [hunk 3]
```diff
@@ -1,14 +1,10 @@
...
    |
 11 | pub mod baz;
    |         ^^^
    |
-note: maybe move this module `$DIR/auxiliary/foo/bar.rs` to its own directory via `$DIR/auxiliary/foo/bar/mod.rs`
-  --> $DIR/auxiliary/foo/bar.rs:11:9
-   |
-11 | pub mod baz;
-   |         ^^^
+   = help: name the file either bar/baz.rs or bar/baz/mod.rs inside the directory "$DIR/auxiliary/foo"
```
Comment (nikomatsakis) [inline src/test/ui/invalid-module-declaration/invalid-module-declaration.stderr]: nice

**Review by nikomatsakis — APPROVED** [review]

**Comment by cramertj** [comment]
```
failures:
---- [ui] ui\invalid-module-declaration\invalid-module-declaration.rs stdout ----
	normalized stderr:
error[E0583]: file not found for module `baz`
  --> $DIR/auxiliary/foo/bar.rs:11:9
   |
11 | pub mod baz;
   |         ^^^
   |
   = help: name the file either bar/baz.rs or bar/baz/mod.rs inside the directory "C:/projects/rust/src/test/ui/invalid-module-declaration/auxiliary/foo"
error: aborting due to previous error
expected stderr:
error[E0583]: file not found for module `baz`
  --> $DIR/auxiliary/foo/bar.rs:11:9
   |
11 | pub mod baz;
   |         ^^^
   |
   = help: name the file either bar/baz.rs or bar/baz/mod.rs inside the directory "$DIR/auxiliary/foo"
error: aborting due to previous error
diff of stderr:
 error[E0583]: file not found for module `baz`
   --> $DIR/auxiliary/foo/bar.rs:11:9
    |
 11 | pub mod baz;
    |         ^^^
    |
-   = help: name the file either bar/baz.rs or bar/baz/mod.rs inside the directory "$DIR/auxiliary/foo"
+   = help: name the file either bar/baz.rs or bar/baz/mod.rs inside the directory "C:/projects/rust/src/test/ui/invalid-module-declaration/auxiliary/foo"
 
 error: aborting due to previous error
 ```
Looks like the `$DIR` paths aren't handled correctly on Windows. Anyone know what's up here? The existing UI tests seem to all use `$DIR` with no problem. Is it maybe something to do with the fact that it's in quotes?

**Comment by kennytm** [comment]
@cramertj The quotes shouldn't matter. However, if the original output is `C:/projects/rust/src/…` instead of `C:\projects\rust\src\…` it won't perform the replacement.

**Comment by cramertj** [comment]
@kennytm Hm... it seems like the slashes are going the right direction, though. It's in MSYS. The diff is this part:

```
-   = help: name the file either bar/baz.rs or bar/baz/mod.rs inside the directory "$DIR/auxiliary/foo"
+   = help: name the file either bar/baz.rs or bar/baz/mod.rs inside the directory "C:/projects/rust/src/test/ui/invalid-module-declaration/auxiliary/foo"
 ```

It looks like the replacement just isn't happening.

**Comment by kennytm** [comment]
@cramertj just to clarify, the original needs to be backslashes.

**Comment by cramertj** [comment]
@kennytm Oh, I understand now. Thanks!

**Comment by cramertj** [comment]
FYI it looks like the way others have solved this is to make separate tests (like mod_file_not_exist_windows.rs) and use `ignore` for all of the unix-y platforms in those, and `ignore-windows` in the main one.

**Comment by cramertj** [comment]
@kennytm fixed.

**Comment by nikomatsakis** [comment]
> FYI it looks like the way others have solved this is to make separate tests (like mod_file_not_exist_windows.rs) and use ignore for all of the unix-y platforms in those, and ignore-windows in the main one.

heh, I mean, maybe we should fix the UI test running harness instead? =) Well, I guess a workaround is fine for now.

File: src/tools/tidy/src/features.rs [hunk 4]
```diff
@@ -91,7 +91,8 @@ pub fn check(path: &Path, bad: &mut bool, quiet: bool) {
 
     super::walk_many(&[&path.join("test/compile-fail"),
                        &path.join("test/compile-fail-fulldeps"),
-                       &path.join("test/parse-fail"),],
+                       &path.join("test/parse-fail"),
+                       &path.join("test/ui"),],
```
Comment (est31) [inline src/tools/tidy/src/features.rs]: Btw as my PR  #46532 which I've done in the meantime (and which was motivated by this PR in fact :p) got merged, you won't need this any more if you rebase.

**Comment by cramertj** [comment]
Breaks rustfmt-- ping @nrc.

**Comment by kennytm** [comment]
Please break `rls` as well. `rls` depends on `rustfmt`.

**Comment by cramertj** [comment]
@kennytm Done.

**Comment by cramertj** [comment]
Failed pretty-printing tests:

```
[01:51:02] failures:
[01:51:02] 
[01:51:02] ---- [pretty] run-pass/non_modrs_mods/non_modrs_mods.rs stdout ----
[01:51:02] 	
[01:51:02] error: pretty-printed source does not typecheck
[01:51:02] status: exit code: 101
[01:51:02] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "-" "-Zno-trans" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/non_modrs_mods/non_modrs_mods.pretty-out" "--target=x86_64-unknown-linux-gnu" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/non_modrs_mods/non_modrs_mods.stage2-x86_64-unknown-linux-gnu.pretty.aux" "-Crpath" "-O" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
[01:51:02] stdout:
[01:51:02] ------------------------------------------
[01:51:02] 
[01:51:02] ------------------------------------------
[01:51:02] stderr:
[01:51:02] ------------------------------------------
[01:51:02] error[E0433]: failed to resolve. Could not find `inner_modrs_mod` in `modrs_mod`
[01:51:02]   --> <anon>:19:16
[01:51:02]    |
[01:51:02] 19 |     modrs_mod::inner_modrs_mod::innest::foo();
[01:51:02]    |                ^^^^^^^^^^^^^^^ Could not find `inner_modrs_mod` in `modrs_mod`
[01:51:02] 
[01:51:02] error[E0433]: failed to resolve. Could not find `inner_foors_mod` in `modrs_mod`
[01:51:02]   --> <anon>:20:16
[01:51:02]    |
[01:51:02] 20 |     modrs_mod::inner_foors_mod::innest::foo();
[01:51:02]    |                ^^^^^^^^^^^^^^^ Could not find `inner_foors_mod` in `modrs_mod`
[01:51:02] 
[01:51:02] error[E0433]: failed to resolve. Could not find `inner_modrs_mod` in `foors_mod`
[01:51:02]   --> <anon>:21:16
[01:51:02]    |
[01:51:02] 21 |     foors_mod::inner_modrs_mod::innest::foo();
[01:51:02]    |                ^^^^^^^^^^^^^^^ Could not find `inner_modrs_mod` in `foors_mod`
[01:51:02] 
[01:51:02] error[E0433]: failed to resolve. Could not find `inner_foors_mod` in `foors_mod`
[01:51:02]   --> <anon>:22:16
[01:51:02]    |
[01:51:02] 22 |     foors_mod::inner_foors_mod::innest::foo();
[01:51:02]    |                ^^^^^^^^^^^^^^^ Could not find `inner_foors_mod` in `foors_mod`
[01:51:02] 
[01:51:02] error[E0433]: failed to resolve. Could not find `inner_modrs_mod` in `attr_mod`
[01:51:02]   --> <anon>:23:15
[01:51:02]    |
[01:51:02] 23 |     attr_mod::inner_modrs_mod::innest::foo();
[01:51:02]    |               ^^^^^^^^^^^^^^^ Could not find `inner_modrs_mod` in `attr_mod`
[01:51:02] 
[01:51:02] error: aborting due to 5 previous errors
[01:51:02] 
[01:51:02] 
[01:51:02] ------------------------------------------
[01:51:02] 
[01:51:02] thread '[pretty] run-pass/non_modrs_mods/non_modrs_mods.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2770:8
[01:51:02] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:51:02] 
[01:51:02] 
[01:51:02] failures:
[01:51:02]     [pretty] run-pass/non_modrs_mods/non_modrs_mods.rs
[01:51:02] 
[01:51:02] test result: [31mFAILED(B[m. 2810 passed; 1 failed; 40 ignored; 0 measured; 0 filtered out
```

**Comment by nikomatsakis** [comment]
@cramertj do you know what might be causing that? (I don't know, not without looking into it)

**Comment by cramertj** [comment]
No, but I'll investigate this afternoon.

**Comment by nikomatsakis** [comment]
After some discussion on IRC -- our hypothesis is that the pretty typeck runner is putting files in the wrong place. Most likely outcome is that we will skip pretty printing for this test, but @cramertj is going to confirm the hypothesis.

**Comment by cramertj** [comment]
Rebased and ignored pretty-printing, as this is issue https://github.com/rust-lang/rust/issues/37195.

**Pull request merged**